lib/Builder/SourceFileRenderer.php

```diff
@@ -29,26 +37,36 @@ public function render(SourceFile $sourceFile, string $contents) : string
 
         $pageParameters = $this->preparePageParameters($sourceFile);
 
-        return $this->twigRenderer->render($template, [
+        $parameters = [
             'page' => $pageParameters,
             'site' => $this->site,
-        ]);
+        ];
+
+        if (isset($parameters['page']['controller'])) {
+            $controllerParameters = $this->controllerExecutor->execute($sourceFile);
+
+            $parameters = array_merge($parameters, $controllerParameters);
+        }
+
+        return $this->twigRenderer->render($template, $parameters);
     }
 
     /**
      * @return mixed[]
      */
     private function preparePageParameters(SourceFile $sourceFile) : array
     {
-        return array_merge($sourceFile->getParameters(), [
+        return array_merge($sourceFile->getParameters()->getAll(), [
             'date' => $sourceFile->getDate(),
         ]);
     }
 
     private function prepareTemplate(SourceFile $sourceFile, string $contents) : string
     {
         if ($sourceFile->isLayoutNeeded()) {
-            if (preg_match_all('/{%\s+block\s+(\w+)\s+%}(.*?){%\s+endblock\s+%}/si', $contents, $matches) === 0) {
+            $hasContentBlock = preg_match_all('/{%\s+block\s+(\w+)\s+%}(.*?){%\s+endblock\s+%}/si', $contents, $matches) === 0;
```

Comment:
Is this variable name correct? From `=== 0` I assume it should rather be `$doesNotHaveContentBlock`.
Also when `$contents === ''`, this regex doesn't need to run (condition below will be false anyway).